.github/workflows/javascript.yml

```diff
@@ -0,0 +1,98 @@
+# This file is part of BenchExec, a framework for reliable benchmarking:
+# https://github.com/sosy-lab/benchexec
+#
+# SPDX-FileCopyrightText: 2023 Dirk Beyer <https://www.sosy-lab.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+name: JavaScript Pipeline
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+  workflow_dispatch:
+
+jobs:
+  lint:
+    name: Perform linting
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v2
+
+      - name: Cache node modules
+        uses: actions/cache@v2
+        with:
+          path: benchexec/tablegenerator/react-table/node_modules
+          key: ${{ runner.os }}-node-modules-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-modules-
```

Comment:
I am wondering about `restore-keys`: As far as I understand the [documentation](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#input-parameters-for-the-cache-action), the current definition does not match anything because we never create a cache with the key `${{ runner.os }}-node-modules-` anywhere, so it would never successfully look it up under this name, right?

It is actually a nice trick to use the hash of `package-lock.json` as a cache key, I like that. But if we cache only the cache directory of npm, we might not even need that and can just reuse the cache always?

Reply:
> I am wondering about `restore-keys`: As far as I understand the [documentation](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#input-parameters-for-the-cache-action), the current definition does not match anything because we never create a cache with the key `${{ runner.os }}-node-modules-` anywhere, so it would never successfully look it up under this name, right?

No. The key with the hashed `package.json` is used to build the artifact for caching, but it can be restored by any of the keys mentioned (that is without the hashed `package-lock.json`), so we do indeed make use of the cache. I think the best would be to cache the `.npm` directory with the `package-lock.json` as key, so we do update the `.npm` content as and when the contents of the lockfile change.

Reply:
Yes, my bad. You are correct here. I don't think the `restore_keys` actually serve a purpose for us here (as we have a fixed runner, and the only thing changing is the `package-lock.json`)! I'd prefer to remove the same to simplify the workflow as much as possible. I also tried to find a way to research a way so that we don't have to duplicate the setting up of node and checking out the repository part of the workflows, but I wasn't able to find something on it (other than using a build matrix with different commands like `npm run lint`, `npm run test` etc., which IMO profoundly decreases the legibality of the codebase.

Reply:
If I understand if correctly, then we do need the hash. Let's say I added a new dependency `XYZ` to the project. Then, in that process, I would update my `package-lock.json` and would re-cache the `.npm` directory. This is essential cause now the newly cached `.npm` directory would also have the compiled files for `XYZ`, and thus during `npm ci` would fetch the same from local storage and not the internet. If we omit the hash, then we are essentially fixing the content of the `.npm` directory to the current state. 

Or am I misinterpreting something here?

Reply:
Hm, maybe I am also wrong.

I would think that a cache works such that in your example, after you add the dependency and run CI once, the cache gets updated such that it now includes the new dependency. Every other CI job after that would use the new cache with the added dependency, but it doesn't hurt if there is more in `.npm` than what the job needs.

This is at least what GitLab's cache does and how I understand the term "cache" in general. Is GitHub different?

...

Ok, [found it](https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache). Indeed, GitHub's cache is a one-time thing. And the next section on that page even says it doesn't use the same cache across branches? This is really weird.

Ok, so we really need the hash, and we need the fallback key as well. Note our `package-lock.json` is in `benchexec/tablegenerator/react-table/`, I think we need to specify this path.

If we run `npm ci` with the correct working directory, does it put the `.npm` in the working directory or in the root of the project? This would also change what we need to cache.

Reply:
Aight. I will restore the fallback keys into the workflow so that we should get the caching benefits. The `.npm` is always created with respect to the root (i.e. at `~./npm`), so once we setup the cache in the next step, no matter the directory we run `npm ci` from, we should use the cached dependencies.